include/boost/url/impl/any_allocator.hpp

```diff
@@ -0,0 +1,190 @@
+    using max_align_t = ::max_align_t;
+#else
+    using max_align_t = std::max_align_t;
+#endif
+
+inline std::size_t ceil_div(std::size_t n, std::size_t d)
+{
+    return (n + d - 1) / d;
+}
+
+template <std::size_t N, class Allocator>
+typename std::enable_if<alignof(max_align_t) <= N, void*>::type
+try_allocate_aligned(const Allocator& a, std::size_t n, std::size_t size, std::size_t align)
+{
+    using aligned_t = max_align_t;
+    using aligned_allocator_t = boost::allocator_rebind_t<Allocator, aligned_t>;
+    return aligned_allocator_t(a).allocate(n * ceil_div(size, align));
+}
+
+template <std::size_t N, class Allocator>
+typename std::enable_if<N < alignof(max_align_t), void*>::type
+try_allocate_aligned(const Allocator& a, std::size_t n, std::size_t size, std::size_t align)
+{
+    if (align <= N)
+    {
+        using aligned_t = typename boost::type_with_alignment<N>::type;
+        using aligned_allocator_t = boost::allocator_rebind_t<Allocator, aligned_t>;
+        return aligned_allocator_t(a).allocate(n * ceil_div(size, align));
+    }
+    else
+    {
+        return try_allocate_aligned<2*N>(a, n, size, align);
+    }
+}
+
+template <class Allocator>
+void* allocate_aligned(const Allocator& a, std::size_t n, std::size_t size, std::size_t align)
+{
+    return try_allocate_aligned<2*alignof(char)>(a, n, size, align);
+}
+
+template <std::size_t N, class Allocator>
+typename std::enable_if<alignof(max_align_t) <= N, void>::type
+try_deallocate_aligned(const Allocator& a, void* p, std::size_t n, std::size_t size, std::size_t align)
+{
+    using aligned_t = max_align_t;
+    using aligned_allocator_t = boost::allocator_rebind_t<Allocator, aligned_t>;
+    aligned_allocator_t(a).deallocate(reinterpret_cast<aligned_t*>(p), n * ceil_div(size, align));
+}
+
+template <std::size_t N, class Allocator>
+typename std::enable_if<N < alignof(max_align_t), void>::type
+try_deallocate_aligned(const Allocator& a, void* p, std::size_t n, std::size_t size, std::size_t align)
+{
+    if (align <= N)
+    {
+        using aligned_t = typename boost::type_with_alignment<N>::type;
+        using aligned_allocator_t = boost::allocator_rebind_t<Allocator, aligned_t>;
+        aligned_allocator_t(a).deallocate(reinterpret_cast<aligned_t*>(p), n * ceil_div(size, align));
+    } else {
+        try_deallocate_aligned<2*N>(a, p, n, size, align);
+    }
+}
+
+template <class Allocator>
+void deallocate_aligned(const Allocator& a, void* p, std::size_t n, std::size_t size, std::size_t align)
+{
+    try_deallocate_aligned<2*alignof(char)>(a, p, n, size, align);
+}
+}
+
+template <class T>
+any_allocator<T>::
+any_allocator() noexcept
+: any_allocator(std::allocator<char>{})
+{
+}
+
+template <class T>
+template <class Allocator>
+any_allocator<T>::
+any_allocator(Allocator const& a) noexcept
+{
+    class impl;
+
+    using char_allocator_type = boost::allocator_rebind_t<Allocator, char>;
+
+    class impl
+            : public detail::any_allocator_base
```

Comment:
this isn't quite right because `any_allocator_base` is not the base of `any_allocator`. What I wrote earlier was this:

```
namespace detail {

struct any_allocator_base
{
    struct base
    {
        ...
    };

    template<class T>
    struct impl : base
    {
        ...
    };
};

} // detail
```

Then derive `any_allocator`

```
class any_allocator : private detail::any_allocator_base
{
    ...
```

This way the nested types are only in scope for the derived class and the names make sense